### ISSUE TITLE
hi ,how do what  and get good result in ffhq_encode

### ISSUE THREAD

**Issue by anthonyyuan** [eladrich/pixel2style2pixel#22]
hi  
   can u give some idea , how to use  
batch_size': '8',
 'board_interval': 50,
 'checkpoint_path': None,
 'dataset_type': 'ffhq_encode',
 'device': 'cuda:0',
 'encoder_type': 'GradualStyleEncoder',
 'exp_dir': '',
 'id_lambda': 0.1,
 'image_interval': 100,
 'input_nc': 3,
 'l2_lambda': 1.0,
 'l2_lambda_crop': 0,
 'label_nc': 0,
 'learn_in_w': False,
 'learning_rate': 0.0001,
 'lpips_lambda': 0.8,
 'lpips_lambda_crop': 0,
 'max_steps': 300000,
 'optim_name': 'ranger',
 'resize_factors': None,
 'save_interval': '10000',
 'start_from_latent_avg': True,
 'stylegan_weights': '',
 'test_batch_size': '6',
 'test_workers': 2,
 'train_decoder': False,
 'val_interval': 2500,
 'w_norm_lambda': 0,
 'workers': '8'}

**Comment by anthonyyuan** [comment]
Here, we use pSp to find the latent code of real images in the latent domain of a pretrained StyleGAN generator. when I use it ,but result is bad

**Comment by yuval-alaluf** [comment]
Hi @anthonyyuan ,
Not sure I understand the question.  
If you are trying to train a new model, it seems like you did not set the path to the pretrained StyleGAN model using `--stylegan_weights`.    
If you are trying to run inference using the pSp model we uploaded, please provide more details as to what you mean by the "result is bad".

**Comment by anthonyyuan** [comment]
![1](https://user-images.githubusercontent.com/3069415/97248026-d1fba880-183b-11eb-9523-74d8fa778430.jpg)
i encoder my image ,i do not controll step .finally I think this result is not good .

**Comment by anthonyyuan** [comment]
from argparse import Namespace
import time
import os
import sys
import pprint
import numpy as np
from PIL import Image
import torch
import torchvision.transforms as transforms

sys.path.append(".")
sys.path.append("..")

from datasets import augmentations
from utils.common import tensor2im, log_input_image
from models.psp import pSp

import argparse

torch.cuda.empty_cache()
experiment_type = 'ffhq_encode'
EXPERIMENT_DATA_ARGS = {
    "ffhq_encode": {
        "model_path": "pretrained_models/psp_ffhq_encode.pt",
        "image_path": "notebooks/images/input_img.jpg",
        "transform": transforms.Compose([
            transforms.Resize((256, 256)),
            transforms.ToTensor(),
            transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])])
    },
    "ffhq_frontalize": {
        "model_path": "pretrained_models/psp_ffhq_frontalization.pt",
        "image_path": "notebooks/images/input_img.jpg",
        "transform": transforms.Compose([
            transforms.Resize((256, 256)),
            transforms.ToTensor(),
            transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])])
    },
    "celebs_sketch_to_face": {
        "model_path": "pretrained_models/psp_celebs_sketch_to_face.pt",
        "image_path": "notebooks/images/input_sketch.jpg",
        "transform": transforms.Compose([
            transforms.Resize((256, 256)),
            transforms.ToTensor()])
    },
    "celebs_seg_to_face": {
        "model_path": "pretrained_models/psp_celebs_seg_to_face.pt",
        "image_path": "notebooks/images/input_mask.png",
        "transform": transforms.Compose([
            transforms.Resize((256, 256)),
            augmentations.ToOneHot(n_classes=19),
            transforms.ToTensor()])
    },
    "celebs_super_resolution": {
        "model_path": "pretrained_models/psp_celebs_super_resolution.pt",
        "image_path": "notebooks/images/input_img.jpg",
        "transform": transforms.Compose([
            transforms.Resize((256, 256)),
            augmentations.BilinearResize(factors=[16]),
            transforms.Resize((256, 256)),
            transforms.ToTensor(),
            transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])])
    },
    "toonify": {
        "model_path": "pretrained_models/psp_ffhq_toonify.pt",
        "image_path": "notebooks/images/input_img.jpg",
        "transform": transforms.Compose([
            transforms.Resize((256, 256)),
            transforms.ToTensor(),
            transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])])
    },
}
experiment_type = 'ffhq_encode'
EXPERIMENT_ARGS = EXPERIMENT_DATA_ARGS[experiment_type]

EXPERIMENT_ARGS = EXPERIMENT_DATA_ARGS[experiment_type]
print(str(EXPERIMENT_ARGS))
print(str(EXPERIMENT_ARGS['model_path']))
print( str(os.path.getsize(EXPERIMENT_ARGS['model_path'])))
if os.path.getsize('pretrained_models/psp_ffhq_encode.pt') < 1000000:
  raise ValueError("Pretrained model was unable to be downlaoded correctly!")


model_path = EXPERIMENT_ARGS['model_path']
ckpt = torch.load(model_path, map_location='cpu')
opts = ckpt['opts']
pprint.pprint(opts)
parser = argparse.ArgumentParser(description='aa')

parser.add_argument('--batch_size', type=str, default='6',help='test')
parser.add_argument('--test_batch_size', type=str, default='6',help='test')
parser.add_argument('--save_interval', type=str, default='10000',help='test')
parser.add_argument('--workers', type=str, default='8',help='test')



test_opts = parser.parse_args()

opts.update(vars(test_opts))
pprint.pprint(opts)

opts['checkpoint_path'] = model_path
if 'learn_in_w' not in opts:
    opts['learn_in_w'] = False
opts= Namespace(**opts)
net = pSp(opts)
net.eval()
net.cuda()
print('Model successfully loaded!')


import matplotlib.pyplot as plt


save_path = "/home/anthony/Downloads/pixel2style2pixel-master/img"
image_paths = [os.path.join(save_path, f) for f in os.listdir(save_path) if f.endswith(".jpg")]
n_images = len(image_paths)  

images = []
n_cols = np.ceil(n_images / 2)
fig = plt.figure(figsize=(20, 4))
for idx, image_path in enumerate(image_paths):
    
    img = Image.open(image_path).convert("RGB")
    images.append(img)
    

img_transforms = EXPERIMENT_ARGS['transform']
transformed_images = [img_transforms(image) for image in images]
def run_on_batch(inputs, net, latent_mask=None):
    if latent_mask is None:
        result_batch = net(inputs.to("cuda").float(), randomize_noise=False)
    else:
        result_batch = []
        for image_idx, input_image in enumerate(inputs):
            # get latent vector to inject into our input image
            vec_to_inject = np.random.randn(1, 512).astype('float32')
            _, latent_to_inject = net(torch.from_numpy(vec_to_inject).to("cuda"),
                                      input_code=True,
                                      return_latents=True)
            # get output image with injected style vector
            res = net(input_image.unsqueeze(0).to("cuda").float(),
                      latent_mask=latent_mask,
                      inject_latent=latent_to_inject)
            result_batch.append(res)
        result_batch = torch.cat(result_batch, dim=0)
    return result_batch
batched_images = torch.stack(transformed_images, dim=0)
with torch.no_grad():
    tic = time.time()
    result_images = run_on_batch(batched_images, net, latent_mask=None)
    toc = time.time()
    print('Inference took {:.4f} seconds.'.format(toc - tic))


couple_results = []
n=0
for original_image, result_image in zip(images, result_images):
    result_image = tensor2im(result_image)
    n=n+1
    res = np.concatenate([np.array(original_image.resize((256, 256))),np.array(result_image.resize((256, 256)))], axis=1)
    #res=np.array(result_image.resize((256, 256)))
    res_im = Image.fromarray(res)
    res_im.save("img2/"+str(n)+".jpg")
torch.cuda.empty_cache()

**Comment by yuval-alaluf** [comment]
I am not sure what you define as "not good". The result seems reasonable to me.

**Comment by yuval-alaluf** [comment]
The result you got seems reasonable.

**Comment by anasiri** [comment]
> Here, we use pSp to find the latent code of real images in the latent domain of a pretrained StyleGAN generator. when I use it ,but result is bad

You need to do alignment and cropping before feeding your image.

**Comment by anasiri** [comment]
> Here, we use pSp to find the latent code of real images in the latent domain of a pretrained StyleGAN generator. when I use it ,but result is bad

You need to align and crop your images properly before feeding them to pSp.

**Comment by yuval-alaluf** [comment]
Yes. @anasiri you are correct that the images need to be aligned before feeding them into pSp.  
I feel like the result @anthonyyuan obtained is correct and quite reasonable (and the input does indeed seem aligned). Therefore, I am closing this issue for now. 
If there are further questions, feel free to reopen this issue or open a new one.
